bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java

```diff
@@ -233,7 +235,27 @@ private void startInstance(Object obj)
     {
         log.debug("Starting %s", obj.getClass().getName());
         LifeCycleMethods methods = methodsMap.get(obj.getClass());
-        for (Method postConstruct : methods.methodsFor(PostConstruct.class)) {
+
+        Collection<Method> postConstructMethods = methods.methodsFor(PostConstruct.class);
+        boolean hasPostConstructMethods = !postConstructMethods.isEmpty();
+        boolean hasPreDestroyMethod = methodsMap.get(obj.getClass()).hasFor(PreDestroy.class);
+
+        if (!hasPostConstructMethods && !hasPreDestroyMethod) {
+            // no need to track in startedInstances or managedInstances
+            return;
+        }
+
+        // Guice can double provision instances (in particular with Optional binding).
+        // Protect against calling post-construct and pre-destroy methods more than once
+        if (startedInstances.putIfAbsent(identityHashCode(obj), obj) != null) {
```

Comment:
Can this be a set instead? do we need to track values?

Reply:
I was concerned about not holding on to the reference. Are Object IDs re-used in the VM? I was worried that an object could get garbage collected and then its Object ID would get re-used resulting in a false positive for the map. Is that possible?

Reply:
I don't think that this is possible

Reply:
The identityHashCode is derived from the memory address of the living object. In theory this could be reused but no two living objects can have the same address (obviously). Since the LifecycleManager lives as long as the objects that are managed by it, these will be garbage collected together so there is no risk that this will break.

Reply:
Per offline conversation I've created `ConcurrentWeakIdentitySet` to manage the started instances. PTAL @wendigo

Reply:
FYI - this bug DB entry (and a few others) explain that a JVM is not required to have unique IDs for objects: https://bugs.java.com/bugdatabase/view_bug?bug_id=JDK-8071434